Catalogue : Hash all views when deciding on image filename

danieldresser-ie opened the pull request:
This avoids a dependency on the default view name which caused generateFileName to fail when outside of a script context.

The most concerning part of this is probably that the tests didn't catch it.  This appears to be due to a difference in how the tests handle setting the default view name ( ImageTestCase.setUp scopes a Context with the default view set ), vs how the UI handles setting the default view name ( a script variable ).  This means that in UI, if you're running something which doesn't inherit the script variables ( like apparently Catalogue::imageReceived ), it doesn't have the default view set, but in the tests, it will have it set, so we missed this bug.

We do seem to be encountering a bunch of these sorts of issues ( I think Ivan has been talking to you about some of the things he's found at IE ).  I do kind of wonder whether there could have been a better way of making sure this got set that wasn't dependent on the UI, but I guess at this point we've probably caught most of the edge cases.

The fix for this specific piece of code goes a little beyond just setting the default view - I hash all views, which would be correct even in the possible future where we start using the catalog with multi-view images.

johnhaddon reviewed (changes requested):
Thanks for the quick fix!

> The fix for this specific piece of code goes a little beyond just setting the default view - I hash all views, which would be correct even in the possible future where we start using the catalog with multi-view images.

Yep - this is definitely the right way to go. Actually, it made me realise that there _are_ ways to get multi-view images into the Catalogue already. You can use the folder button to load an image from disk, or you can drag and drop an ImagePlug into the image list. The latter is fixed by this PR. But both result in a subsequent error, which is displayed in the status column in the Catalogue - it seems something else is querying the image with an invalid view (possibly the code that queries data for the columns themselves?). We should fix that too.

![image](https://user-images.githubusercontent.com/1133871/178941949-f1175443-5c98-4b98-a2b7-c085b3fcd3d7.png)

> The most concerning part of this is probably that the tests didn't catch it.

Probably worth adding a test that calls `generateFileName( aMultiViewImagePlug )`? Although actually, I just realised that the most concerning thing is that this bug was preventing completed renders from being saved in the catalogue at all. We definitely need to call this out with an entry in Changes.md.

> We do seem to be encountering a bunch of these sorts of issues ( I think Ivan has been talking to you about some of the things he's found at IE ). I do kind of wonder whether there could have been a better way of making sure this got set that wasn't dependent on the UI, but I guess at this point we've probably caught most of the edge cases.

Yeah, Ivan has brought this up with me. I think there's a case for a better mechanism for defaulting `image:viewName` - maybe that default context we were talking about before. But I'm hoping we can hobble on this way a bit longer.

We also need to retarget this PR to `1.0_maintenance`, so that gives us :

- [ ] Fix error triggered by loading a multi-view image
- [ ] Update Changes.md
- [ ] Retarget to `1.0_maintenance`.

Cheers...
John

danieldresser-ie commented:
Addressed feedback ( along with a couple of other minor things that came up along the way ).  I think this is all good to go.

During testing though, I did come across one exceptionally unfortunate thing that I'm sorry I didn't catch sooner.  I'd tested all of the stuff that might be hard to vary between views, but because metadata is stored separately, I guess I just assumed it would be fine.

Turns out that there's this function in OpenEXR checkSharedAttributesValues : https://github.com/AcademySoftwareFoundation/openexr/blob/2205245599709d88013b7da03b176c1c73c8aedb/src/lib/OpenEXR/ImfMultiPartOutputFile.cpp#L383

This prevents you writing EXRs with differing metadata, if the metadata is one of: displayWindow, pixelAspect, timeCode or chromaticity.  I guess this restriction makes some sense when using multiview for stereo, but it could throw a real wrench in some of our plans for using multi-view for generic networks that process multiple separate images through the same network.  I suppose that for most of our plans for using multi-view for processing multiple independent images, they wouldn't be written to the same file, so hopefully it's not that big an annoyance ... but when the format is technically completely capable of storing this stuff separately, it's a shame not to be able to turn off this check.

johnhaddon commented on python/GafferImageUI/CatalogueUI.py:
```diff
@@ -210,7 +219,7 @@ def _imageCellData( self, image, catalogue ) :
 		value = self.__defaultValue
 
-		metadata = catalogue["out"].metadata()
+		metadata = catalogue["out"].metadata( self._standardViewName( catalogue["out"] ) )
```
Rather than fix this manually for every single column (other people have custom columns, and there are columns in Gaffer that aren't fixed in this commit - see `startup/GafferSceneUI/catalogue.py`), we should just set `image:viewName` once at line 109, before delegating to `_imageCellData`.

danieldresser-ie commented:
Oh, yeah, that's a lot cleaner after moving the context var to before delegating to _imageCellData.  Squashed the fix in.
